crates/revm/src/context.rs

```diff
@@ -1,30 +1,58 @@
 use crate::{
-    db::Database,
+    db::{Database, EmptyDB},
     interpreter::{
         analysis::to_analysed, gas, return_ok, CallInputs, Contract, CreateInputs, Gas,
         InstructionResult, Interpreter, InterpreterResult, MAX_CODE_SIZE,
     },
     journaled_state::JournaledState,
     precompile::{Precompile, Precompiles},
     primitives::{
-        keccak256, Address, AnalysisKind, Bytecode, Bytes, EVMError, Env, Spec, SpecId::*, B256,
-        U256,
+        keccak256, Address, AnalysisKind, Bytecode, Bytes, EVMError, Env, HashSet, Spec, SpecId,
+        SpecId::*, B256, U256,
     },
-    CallStackFrame, CALL_STACK_LIMIT,
+    CallStackFrame, FrameOrResult, CALL_STACK_LIMIT,
 };
 use alloc::boxed::Box;
 use core::ops::Range;
 
-/// EVM Data contains all the data that EVM needs to execute.
+/// Main Context structure that contains both EvmContext and external contexts.
+pub struct Context<EXT, DB: Database> {
+    /// Evm Context.
+    pub evm: EvmContext<DB>,
+    /// External contexts.
+    pub external: EXT,
+}
+
+impl<EXT, DB: Database> Context<EXT, DB> {
+    /// Creates empty context. This is useful for testing.
+    pub fn new_empty() -> Context<(), EmptyDB> {
+        Context {
+            evm: EvmContext::new(EmptyDB::new()),
+            external: (),
+        }
+    }
+}
+
+impl<DB: Database> Context<(), DB> {
+    /// Creates new context with database.
+    pub fn new_with_db(db: DB) -> Context<(), DB> {
+        Context {
+            evm: EvmContext::new_with_env(db, Box::new(Env::default())),
+            external: (),
+        }
+    }
+}
+
+/// EVM contexts contains data that EVM needs for execution.
 #[derive(Debug)]
-pub struct EvmContext<'a, DB: Database> {
+pub struct EvmContext<DB: Database> {
```

Comment:
I had a look at the PR to determine whether these changes would also resolve the lifetime issue that we reported in https://github.com/bluealloy/revm/pull/456/

Unfortunately, the presence of the `DB` generic in `EvmContext` makes it such that that generic type parameter also bleeds into the `Inspector` trait, thus tying the lifetime of `DB` to the `Inspector`.

This happens if `DB` is a reference type `&SomeDatabase`; a common pattern when the database is owned by someone else and is only temporarily passed into the `Evm` to run a transaction.

One way to resolve this would be the proposed solution in https://github.com/bluealloy/revm/pull/456/: The inspector receives a reference to a dynamic trait object `EvmContext` that has a generic type `DatabaseErrorT` instead of the `Database` generic type.

I'm happy to experiment with any solutions you have in mind or point you to a branch of EDR to experiment on

Reply:
Nobody stops you from defining your version of Inspector and your handler register. You can even define the needed trait outside of revm and implement it on the EvmContext ([this trait](https://github.com/bluealloy/revm/pull/456/files#diff-43d7eb2000c0f76dbb285beb1d5aa9742c5ec725db8db4914fe4d6ffc06e7fcbR220-R233)) and you is in your own inspector.

This PR cuts the connection between Inspector and revm, and makes Inspector customisable in a way that anything can be used in its place.

Reply:
> This PR cuts the connection between Inspector and revm, and makes Inspector customisable in a way that anything can be used in its place.

Apologies, I missed that part. I thought they were still tied together.

I'll try defining my own inspector type to see whether that resolves the issue

Reply:
One more observation (tbh not sure if it is valid) but the lifetime is removed from `EvmContext` so the `Inspector` just takes the Database there is no lifetime restriction.

so doesn't this mean you can specify the `Inspector` with db reference as.

```
impl<'a> Inspector<&'a DbType> for InspectorImpl {

}
```

`InspectorImpl` doesn't neet to have reference